votor/src/certificate_pool/slot_stake_counters.rs

```diff
@@ -0,0 +1,301 @@
+use {
+    crate::{
+        certificate_pool::stats::CertificatePoolStats, event::VotorEvent, Stake,
+        SAFE_TO_NOTAR_MIN_NOTARIZE_AND_SKIP, SAFE_TO_NOTAR_MIN_NOTARIZE_FOR_NOTARIZE_OR_SKIP,
+        SAFE_TO_NOTAR_MIN_NOTARIZE_ONLY, SAFE_TO_SKIP_THRESHOLD,
+    },
+    solana_sdk::hash::Hash,
+    solana_votor_messages::vote::Vote,
+    std::collections::BTreeMap,
+};
+
+#[derive(Debug, Default)]
+pub(crate) struct SlotStakeCounters {
+    my_first_vote: Option<Vote>,
+    total_stake: Stake,
+    skip_total: Stake,
+    notarize_total: Stake,
+    notarize_entry_total: BTreeMap<Hash, Stake>,
+    top_notarized_stake: Stake,
+    safe_to_notar_sent: Vec<Hash>,
+    safe_to_skip_sent: bool,
+}
+
+impl SlotStakeCounters {
+    pub fn add_vote(
+        &mut self,
+        vote: &Vote,
+        total_stake: Stake,
+        entry_stake: Stake,
+        is_my_own_vote: bool,
+        events: &mut Vec<VotorEvent>,
+        stats: &mut CertificatePoolStats,
+    ) {
+        match vote {
+            Vote::Skip(_) => self.skip_total = entry_stake,
+            Vote::Notarize(vote) => {
+                self.notarize_entry_total
+                    .insert(*vote.block_id(), entry_stake);
+                self.notarize_total = self.notarize_entry_total.values().sum();
+                self.top_notarized_stake = self.top_notarized_stake.max(entry_stake);
+            }
+            _ => return, // Not interested in other vote types
+        }
+        if self.my_first_vote.is_none() && is_my_own_vote {
+            self.my_first_vote = Some(*vote);
+            self.total_stake = total_stake;
+        }
+        if self.my_first_vote.is_some() {
+            let slot = vote.slot();
+            // Check safe to notar
+            for (block_id, stake) in &self.notarize_entry_total {
+                if !self.safe_to_notar_sent.contains(block_id)
+                    && self.is_safe_to_notar(block_id, stake)
+                {
+                    events.push(VotorEvent::SafeToNotar((slot, *block_id)));
+                    stats.event_safe_to_notarize = stats.event_safe_to_notarize.saturating_add(1);
+                    self.safe_to_notar_sent.push(*block_id);
+                }
+            }
+            // Check safe to skip
+            if !self.safe_to_skip_sent && self.is_safe_to_skip() {
+                events.push(VotorEvent::SafeToSkip(slot));
+                self.safe_to_skip_sent = true;
+                stats.event_safe_to_skip = stats.event_safe_to_skip.saturating_add(1);
+            }
+        }
+    }
+
+    fn is_safe_to_notar(&self, block_id: &Hash, stake: &Stake) -> bool {
+        // White paper v1.1 page 22: The event is only issued if the node voted in slot s already,
+        // but not to notarize b. Moreover:
+        // notar(b) >= 40% or (skip(s) + notar(b) >= 60% and notar(b) >= 20%)
+        if let Some(Vote::Notarize(my_vote)) = self.my_first_vote.as_ref() {
+            if my_vote.block_id() == block_id {
```

Comment:
Do we also want to avoid revoting on second and third blocks we have already voted on

Reply:
Only Notarize count here, there can only be one Notarize. (NotarizeFallback doesn't count here.)

Reply:
ah yeah, makes sense